<a href="https://colab.research.google.com/github/theMorana/HSE/blob/main/Machine%20Learning/Semester_2_Homeworks%20/ML_Hometask_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [3]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [2]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

depth=3: acc=0.6244, time=0.17s
depth=5: acc=0.6971, time=0.28s
depth=10: acc=0.7716, time=0.37s
depth=20: acc=0.8409, time=0.48s
depth=None: acc=0.8697, time=0.65s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.17 |
|           5 |          0.6971 |       0.28 |
|          10 |          0.7716 |       0.37 |
|          20 |          0.8409 |       0.48 |
|         nan |          0.8697 |       0.65 |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [3]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

# Ваш код здесь (перебор max_depth)\

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(max_depth=depth, n_estimators=100, random_state=42, n_jobs=-1)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

depth=3: acc=0.8799, time=0.40s
depth=5: acc=0.9002, time=0.39s
depth=10: acc=0.9154, time=0.54s
depth=20: acc=0.9357, time=0.85s
depth=None: acc=0.9526, time=1.54s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.8799 |       0.4  |
|           5 |          0.9002 |       0.39 |
|          10 |          0.9154 |       0.54 |
|          20 |          0.9357 |       0.85 |
|         nan |          0.9526 |       1.54 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [4]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(learning_rate=lr, n_estimators=100, max_depth=3, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({'learning_rate': lr,'test_accuracy': round(acc, 4),'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")



df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=36.53s
learning_rate=0.05: acc=0.9475, time=47.11s
learning_rate=0.1: acc=0.9577, time=36.23s
learning_rate=0.5: acc=0.9662, time=35.82s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      36.53 |
|            0.05 |          0.9475 |      47.11 |
|            0.1  |          0.9577 |      36.23 |
|            0.5  |          0.9662 |      35.82 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [6]:
from xgboost import XGBClassifier

results_xgb = []

# Ваш код здесь (перебор gamma, use_label_encoder=False, eval_metric='logloss')
for gm in [0, 0.1, 0.5, 1.0]:
    start = time.time()
    clf = XGBClassifier(gamma=gm, n_estimators=100, learning_rate=0.1, eval_metric='logloss', random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma':gm,'test_accuracy': round(acc, 4),'time_sec': round(fit_time, 2)})
    print(f"gamma={gm}: acc={acc:.4f}, time={fit_time:.2f}s")



df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))

gamma=0: acc=0.9459, time=40.01s
gamma=0.1: acc=0.9442, time=32.88s
gamma=0.5: acc=0.9492, time=32.62s
gamma=1.0: acc=0.9492, time=29.58s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      40.01 |
|     0.1 |          0.9442 |      32.88 |
|     0.5 |          0.9492 |      32.62 |
|     1   |          0.9492 |      29.58 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [7]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

# Ваш код здесь (перебор learning_rate)
for lr in [0.5, 1.0, 1.5, 2.0]:
    start = time.time()
    clf = AdaBoostClassifier(learning_rate=lr, n_estimators=100, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({'learning_rate': lr,'test_accuracy': round(acc, 4),'time_sec': round(fit_time, 2)})
    print(f"learning_rate={lr}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=5.26s
learning_rate=1.0: acc=0.9205, time=5.91s
learning_rate=1.5: acc=0.9255, time=5.30s
learning_rate=2.0: acc=0.8731, time=5.55s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |       5.26 |
|             1   |          0.9205 |       5.91 |
|             1.5 |          0.9255 |       5.3  |
|             2   |          0.8731 |       5.55 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [8]:
from lightgbm import LGBMClassifier

results_lgbm = []

# Ваш код здесь (перебор num_leaves, n_jobs=-1)
for num_leaves_val in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(num_leaves=num_leaves_val, n_estimators=100, learning_rate=0.1, n_jobs=-1, random_state=42, verbose=-1)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({'num_leaves': num_leaves_val, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"num_leaves={num_leaves_val}: acc={acc:.4f}, time={fit_time:.2f}s")

df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=15: acc=0.9662, time=6.51s


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=31: acc=0.9594, time=7.57s


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=63: acc=0.9577, time=13.42s
num_leaves=127: acc=0.9594, time=20.17s

Таблица 6. LightGBM
|   num_leaves |   test_accuracy |   time_sec |
|-------------:|----------------:|-----------:|
|           15 |          0.9662 |       6.51 |
|           31 |          0.9594 |       7.57 |
|           63 |          0.9577 |      13.42 |
|          127 |          0.9594 |      20.17 |


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [9]:
!pip install xgboost lightgbm catboost -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 8.6 MB/s eta 0:00:00


In [10]:
from catboost import CatBoostClassifier

results_cat = []

# Ваш код здесь (перебор depth, verbose=0)
for depth_val in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(depth=depth_val, iterations=100, learning_rate=0.1, verbose=0, random_seed=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({'depth': depth_val, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth_val}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

depth=3: acc=0.9340, time=41.84s
depth=5: acc=0.9425, time=39.48s
depth=7: acc=0.9442, time=122.31s
depth=10: acc=0.9526, time=773.73s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |      41.84 |
|       5 |          0.9425 |      39.48 |
|       7 |          0.9442 |     122.31 |
|      10 |          0.9526 |     773.73 |


## 4. Итоговая таблица лучших результатов

In [4]:
# Соберите лучшие результаты из каждой таблицы вручную или автоматически
# Пример структуры:

summary = pd.DataFrame([
    ('Decision Tree', 'nan', 0.8697, 0.65),  # замените 0 на ваши значения
    ('Random Forest', 'nan', 0.95, 1.54),
    ('Gradient Boosting', 0.5, 0.9662, 35.82),
    ('XGBoost', 1, 0.9492, 29.58),
    ('AdaBoost', 1.5, 0.9255, 5.3),
    ('LightGBM', 15, 0.9662, 6.51),
    ('CatBoost', 5, 0.9425, 39.48), #я выбрала именно этот параметр, потому что самая высокая accuracy занимает
    #в три раза больше времени при увеличении качества всего на 17 сотых
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print("Итоговая таблица лучших результатов")
print(summary.to_markdown(index=False))


Итоговая таблица лучших результатов
| Algorithm         |   Best params |   Best test acc |   Time (sec) |
|:------------------|--------------:|----------------:|-------------:|
| Decision Tree     |         nan   |          0.8697 |         0.65 |
| Random Forest     |         nan   |          0.95   |         1.54 |
| Gradient Boosting |           0.5 |          0.9662 |        35.82 |
| XGBoost           |           1   |          0.9492 |        29.58 |
| AdaBoost          |           1.5 |          0.9255 |         5.3  |
| LightGBM          |          15   |          0.9662 |         6.51 |
| CatBoost          |           5   |          0.9425 |        39.48 |


## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:** Максимальная точность достигнута алгоритмами Gradient Boosting и LightGBM. Их параметры: learning_rate=0.5 и num_leaves=15.


**Ответ 2:** Согласно итоговой таблице, самый быстрый алгоритм (Decision Tree = 0.65s) обучался быстрее самого медленного (CatBoost = 39.48 s) на 38.83(в 57 раз).


**Самое быстрое** обучение среди всех - Decision Tree
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.17 |


**Cамое медленное** обучение среди всех - Cat Boost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       7 |          0.9442 |     122.31 |

Разница составляет 719 %


**Ответ 3:** Переобучение наблюдается у Decision Tree (при gamma = none). Дерево растет до тех пор, пока все листья не станут чистыми (100% accuracy на обучении). Также небольшое переобучение есть у Random Forest (max_depth=None) и XGBoost (gamma=0), но оно не так ярко выражено.


**Ответ 4:** Я бы выбрала LightGBM, так как у этого алгоритма самые высокие показатели в соотношении качества / времени выполнения. В продакшн-системах нужно быстро обрабатывать большой объём текстов без потери качества, так как время может значительно влиять на обработку данных системой, а большой объём текстов требует высокого качества.